File: docs/api/examples/python/console_client_example.py
"""
Takhin Console REST API Client - Python Example

This example demonstrates how to interact with the Takhin Console REST API
using the Python requests library.

Install dependencies:
    pip install requests
"""

import requests
from typing import List, Dict, Optional
import json


class TakhinClient:
    """Client for Takhin Console REST API"""

    def __init__(self, base_url: str = "http://localhost:8080/api", api_key: Optional[str] = None):
        """
        Initialize the client
        
        Args:
            base_url: Base URL of the Console API
            api_key: Optional API key for authentication
        """
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
        if api_key:
            self.session.headers["Authorization"] = f"Bearer {api_key}"

    def _request(self, method: str, path: str, **kwargs) -> requests.Response:
        """Make an HTTP request"""
        url = f"{self.base_url}{path}"
        resp = self.session.request(method, url, **kwargs)
        resp.raise_for_status()
        return resp

    # Health endpoints

    def health(self) -> Dict:
        """Get health status"""
        resp = self._request("GET", "/health")
        return resp.json()

    def health_ready(self) -> Dict:
        """Get readiness status"""
        resp = self._request("GET", "/health/ready")
        return resp.json()

    # Topic operations

    def list_topics(self) -> List[Dict]:
        """List all topics"""
        resp = self._request("GET", "/topics")
        return resp.json()

    def get_topic(self, name: str) -> Dict:
        """Get topic details"""
        resp = self._request("GET", f"/topics/{name}")
        return resp.json()

    def create_topic(self, name: str, partitions: int) -> Dict:
        """Create a new topic"""
        data = {"name": name, "partitions": partitions}
        resp = self._request("POST", "/topics", json=data)
        return resp.json()

    def delete_topic(self, name: str) -> None:
        """Delete a topic"""
        self._request("DELETE", f"/topics/{name}")

    # Message operations

    def get_messages(
        self,
        topic: str,
        partition: int,
        offset: int,
        limit: int = 100
    ) -> List[Dict]:
        """
        Get messages from a topic partition
        
        Args:
            topic: Topic name
            partition: Partition ID
            offset: Starting offset
            limit: Maximum number of messages to return
        """
        params = {
            "partition": partition,
            "offset": offset,
            "limit": limit
        }
        resp = self._request("GET", f"/topics/{topic}/messages", params=params)
        return resp.json()

    def produce_message(
        self,
        topic: str,
        partition: int,
        key: str,
        value: str
    ) -> Dict:
        """
        Produce a message to a topic
        
        Args:
            topic: Topic name
            partition: Partition ID
            key: Message key
            value: Message value
        """
        data = {
            "partition": partition,
            "key": key,
            "value": value
        }
        resp = self._request("POST", f"/topics/{topic}/messages", json=data)
        return resp.json()

    # Consumer group operations

    def list_consumer_groups(self) -> List[Dict]:
        """List all consumer groups"""
        resp = self._request("GET", "/consumer-groups")
        return resp.json()

    def get_consumer_group(self, group_id: str) -> Dict:
        """Get consumer group details"""
        resp = self._request("GET", f"/consumer-groups/{group_id}")
        return resp.json()


def main():
    """Example usage"""
    
    # Initialize client (set api_key if authentication is enabled)
    client = TakhinClient(api_key=None)

    print("=== Health Check ===")
    health = client.health()
    print(f"Status: {health['status']}\n")

    # Create topic
    print("=== Create Topic ===")
    topic_name = "events"
    try:
        result = client.create_topic(topic_name, 3)
        print(f"Topic '{topic_name}' created: {result}")
    except requests.HTTPError as e:
        print(f"Warning: {e}")
    print()

    # List topics
    print("=== List Topics ===")
    topics = client.list_topics()
    for topic in topics:
        print(f"Topic: {topic['name']}, Partitions: {topic['partitionCount']}")
        for partition in topic.get('partitions', []):
            print(f"  Partition {partition['id']}: HWM={partition['highWaterMark']}")
    print()

    # Produce messages
    print("=== Produce Messages ===")
    messages = [
        {"key": "user-1", "value": json.dumps({"action": "login", "timestamp": "2024-01-01T10:00:00Z"})},
        {"key": "user-2", "value": json.dumps({"action": "purchase", "amount": 99.99})},
        {"key": "user-1", "value": json.dumps({"action": "logout", "timestamp": "2024-01-01T11:00:00Z"})},
    ]

    for msg in messages:
        try:
            result = client.produce_message(topic_name, 0, msg["key"], msg["value"])
            print(f"Produced: key={msg['key']}, offset={result['offset']}, partition={result['partition']}")
        except requests.HTTPError as e:
            print(f"Failed to produce: {e}")
    print()

    # Read messages
    print("=== Read Messages ===")
    read_messages = client.get_messages(topic_name, partition=0, offset=0, limit=10)
    for msg in read_messages:
        print(f"Offset {msg['offset']}: key={msg['key']}, value={msg['value']}")
    print()

    # List consumer groups
    print("=== List Consumer Groups ===")
    groups = client.list_consumer_groups()
    for group in groups:
        print(f"Group: {group['groupId']}, State: {group['state']}, Members: {group['members']}")
    print()

    # Get consumer group details
    if groups:
        print("=== Consumer Group Details ===")
        group_id = groups[0]['groupId']
        details = client.get_consumer_group(group_id)
        print(f"Group: {details['groupId']}")
        print(f"State: {details['state']}")
        print(f"Protocol: {details.get('protocol', 'N/A')}")
        print("Members:")
        for member in details.get('members', []):
            print(f"  {member['memberId']} ({member['clientId']})")
        print("Offset Commits:")
        for commit in details.get('offsetCommits', []):
            print(f"  {commit['topic']}[{commit['partition']}] = {commit['offset']}")
        print()

    print("All examples completed!")


if __name__ == "__main__":
    main()

File: docs/api/examples/javascript/console_client_example.ts
/**
 * Takhin Console REST API Client - JavaScript/TypeScript Example
 * 
 * This example demonstrates how to interact with the Takhin Console REST API
 * using the Fetch API (available in Node.js 18+ and browsers).
 * 
 * For Node.js < 18, install node-fetch:
 *   npm install node-fetch
 */

const BASE_URL = 'http://localhost:8080/api';
const API_KEY = null; // Set if authentication is enabled

interface CreateTopicRequest {
  name: string;
  partitions: number;
}

interface Topic {
  name: string;
  partitionCount: number;
  partitions: Partition[];
}

interface Partition {
  id: number;
  highWaterMark: number;
}

interface ProduceMessageRequest {
  partition: number;
  key: string;
  value: string;
}

interface ProduceMessageResponse {
  offset: number;
  partition: number;
}

interface Message {
  partition: number;
  offset: number;
  key: string;
  value: string;
  timestamp: number;
}

interface ConsumerGroup {
  groupId: string;
  state: string;
  members: number;
}

interface ConsumerGroupDetails {
  groupId: string;
  state: string;
  protocolType: string;
  protocol: string;
  members: GroupMember[];
  offsetCommits: OffsetCommitEntry[];
}

interface GroupMember {
  memberId: string;
  clientId: string;
  clientHost: string;
  partitions: PartitionAssignment[];
}

interface PartitionAssignment {
  topic: string;
  partition: number;
}

interface OffsetCommitEntry {
  topic: string;
  partition: number;
  offset: number;
  metadata: string;
}

class TakhinClient {
  private baseURL: string;
  private apiKey: string | null;

  constructor(baseURL: string = BASE_URL, apiKey: string | null = API_KEY) {
    this.baseURL = baseURL;
    this.apiKey = apiKey;
  }

  private async request<T>(
    method: string,
    path: string,
    body?: any
  ): Promise<T> {
    const headers: Record<string, string> = {};
    
    if (body) {
      headers['Content-Type'] = 'application/json';
    }
    
    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const response = await fetch(`${this.baseURL}${path}`, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`HTTP ${response.status}: ${error}`);
    }

    if (response.status === 204) {
      return null as T;
    }

    return response.json();
  }

  // Health endpoints
  
  async health(): Promise<{ status: string }> {
    return this.request('GET', '/health');
  }

  async healthReady(): Promise<any> {
    return this.request('GET', '/health/ready');
  }

  // Topic operations

  async listTopics(): Promise<Topic[]> {
    return this.request('GET', '/topics');
  }

  async getTopic(name: string): Promise<Topic> {
    return this.request('GET', `/topics/${name}`);
  }

  async createTopic(name: string, partitions: number): Promise<void> {
    await this.request('POST', '/topics', { name, partitions });
  }

  async deleteTopic(name: string): Promise<void> {
    await this.request('DELETE', `/topics/${name}`);
  }

  // Message operations

  async getMessages(
    topic: string,
    partition: number,
    offset: number,
    limit: number = 100
  ): Promise<Message[]> {
    const params = new URLSearchParams({
      partition: partition.toString(),
      offset: offset.toString(),
      limit: limit.toString(),
    });
    return this.request('GET', `/topics/${topic}/messages?${params}`);
  }

  async produceMessage(
    topic: string,
    partition: number,
    key: string,
    value: string
  ): Promise<ProduceMessageResponse> {
    return this.request('POST', `/topics/${topic}/messages`, {
      partition,
      key,
      value,
    });
  }

  // Consumer group operations

  async listConsumerGroups(): Promise<ConsumerGroup[]> {
    return this.request('GET', '/consumer-groups');
  }

  async getConsumerGroup(groupId: string): Promise<ConsumerGroupDetails> {
    return this.request('GET', `/consumer-groups/${groupId}`);
  }
}

// Example usage
async function main() {
  const client = new TakhinClient();

  try {
    // Health check
    console.log('=== Health Check ===');
    const health = await client.health();
    console.log(`Status: ${health.status}\n`);

    // Create topic
    console.log('=== Create Topic ===');
    const topicName = 'events';
    try {
      await client.createTopic(topicName, 3);
      console.log(`Topic '${topicName}' created`);
    } catch (error) {
      console.log(`Warning: ${error}`);
    }
    console.log();

    // List topics
    console.log('=== List Topics ===');
    const topics = await client.listTopics();
    topics.forEach(topic => {
      console.log(`Topic: ${topic.name}, Partitions: ${topic.partitionCount}`);
      topic.partitions.forEach(p => {
        console.log(`  Partition ${p.id}: HWM=${p.highWaterMark}`);
      });
    });
    console.log();

    // Produce messages
    console.log('=== Produce Messages ===');
    const messages = [
      { key: 'user-1', value: JSON.stringify({ action: 'login', timestamp: '2024-01-01T10:00:00Z' }) },
      { key: 'user-2', value: JSON.stringify({ action: 'purchase', amount: 99.99 }) },
      { key: 'user-1', value: JSON.stringify({ action: 'logout', timestamp: '2024-01-01T11:00:00Z' }) },
    ];

    for (const msg of messages) {
      const result = await client.produceMessage(topicName, 0, msg.key, msg.value);
      console.log(`Produced: key=${msg.key}, offset=${result.offset}, partition=${result.partition}`);
    }
    console.log();

    // Read messages
    console.log('=== Read Messages ===');
    const readMessages = await client.getMessages(topicName, 0, 0, 10);
    readMessages.forEach(msg => {
      console.log(`Offset ${msg.offset}: key=${msg.key}, value=${msg.value}`);
    });
    console.log();

    // List consumer groups
    console.log('=== List Consumer Groups ===');
    const groups = await client.listConsumerGroups();
    groups.forEach(group => {
      console.log(`Group: ${group.groupId}, State: ${group.state}, Members: ${group.members}`);
    });
    console.log();

    // Get consumer group details
    if (groups.length > 0) {
      console.log('=== Consumer Group Details ===');
      const groupId = groups[0].groupId;
      const details = await client.getConsumerGroup(groupId);
      console.log(`Group: ${details.groupId}`);
      console.log(`State: ${details.state}`);
      console.log(`Protocol: ${details.protocol}`);
      console.log('Members:');
      details.members.forEach(member => {
        console.log(`  ${member.memberId} (${member.clientId})`);
      });
      console.log('Offset Commits:');
      details.offsetCommits.forEach(commit => {
        console.log(`  ${commit.topic}[${commit.partition}] = ${commit.offset}`);
      });
      console.log();
    }

    console.log('All examples completed!');
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

// Run examples
if (require.main === module) {
  main();
}

export { TakhinClient };
